common/common.go

```diff
@@ -199,6 +199,8 @@ const (
 	EnvControllerReplicas = "ARGOCD_CONTROLLER_REPLICAS"
 	// EnvControllerShard is the shard number that should be handled by controller
 	EnvControllerShard = "ARGOCD_CONTROLLER_SHARD"
+	// EnvControllerShardingAlgorithm is the distribution sharding algorithm to be used: legacy or hash
+	EnvControllerShardingAlgorithm = "ARGOCD_CONTROLLER_SHARDING_ALGORITHM"
```

Comment:
Why do we want to choose the algorithm? What are the pros and cons of each? If one is clearly "better", why do we want to keep the legacy sharding algorithm?

Reply:
Here are some reasons for that:
- maintain the possibility for users who were using the old placement algorithm to stick to it in case they were basing some of their workflows or if they have any workaround. Once, the new algorithm has been proven to be safe, performant and scalable enough, it can become the default.
- give the ability to developers  to implement other alogorithm that could be adapted to other situations. In some scenario, simpler algortihms or more elaborated can be interesting. We were thinking to have a Random distribution algorithm, which does not produce any overhead by querying the cluster list and, for a large enough number of clusters, it produces satfisfiable results
- it is opening the door to more elaborated and dynamic implementations. Right now, to ensure that the new number of replicas is taken into account, a restart of all the pods is required. We want to work safely on other implementation where this could be more dynamic. So, introducing this new api will help on that, by not breaking any existing configuration.


-

Reply:
@akram that makes sense, thanks for the explanation!